data/manager.php

```diff
@@ -0,0 +1,214 @@
+	 */
+	public function get_controllers() {
+		return $this->controllers;
+	}
+
+	/**
+	 * Register controller.
+	 *
+	 * @param string $controller_class_name
+	 */
+	private function register_controller( $controller_class_name ) {
+		$controller_instance = new $controller_class_name();
+
+		// TODO: Validate instance.
+
+		$this->controllers[ $controller_instance->get_name() ] = $controller_instance;
+	}
+
+	/**
+	 * Find controller instance.
+	 *
+	 * By given command name.
+	 *
+	 * @param string $command
+	 *
+	 * @return false|\Elementor\Data\Base\Controller
+	 */
+	public function find_controller_instance( $command ) {
+		$command_parts = explode( '/', $command );
+		$assumed_command_parts = [];
+
+		foreach ( $command_parts as $command_part ) {
+			$assumed_command_parts [] = $command_part;
+
+			foreach ( $this->controllers as $controller_name => $controller ) {
+				$assumed_command = implode( '/', $assumed_command_parts );
+
+				if ( $assumed_command === $controller_name ) {
+					return $controller;
+				}
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Command to endpoint.
+	 *
+	 * Format is required otherwise $command will returned.
+	 *
+	 * @param string $command
+	 * @param string $format
+	 * @param array   $args
+	 *
+	 * @return string endpoint
+	 */
+	public function command_to_endpoint( $command, $format, $args ) {
+		$endpoint = $command;
+
+		if ( $format ) {
+			$formatted = $format;
+
+			array_walk( $args, function ( $val, $key ) use ( &$formatted ) {
+				$formatted = str_replace( ':' . $key, $val, $formatted );
+			} );
+
+			// Remove if not requested.
+			if ( strstr( $formatted, '/:' ) ) {
+				$formatted = substr( $formatted, 0, strpos( $formatted, '/:' ) );
+			}
+
+			$endpoint = $formatted;
+		}
+
+		return $endpoint;
+	}
+
+	/**
+	 * Run processor.
+	 *
+	 * @param \Elementor\Data\Base\Processor $processor
+	 * @param array                          $data
+	 *
+	 * @return mixed
+	 */
+	public static function run_processor( $processor, $data ) {
+		if ( call_user_func_array( [ $processor, 'get_conditions' ], $data ) ) {
+			return call_user_func_array( [ $processor, 'apply' ], $data );
+		}
+
+		return null;
+	}
+
+	/**
+	 * Run processors.
+	 *
+	 * Filter them by class.
+	 *
+	 * @param \Elementor\Data\Base\Processor[] $processors
+	 * @param string                           $filter_by_class
+	 * @param array                            $data
+	 *
+	 * @return false|array
+	 */
+	public static function run_processors( $processors, $filter_by_class, $data ) {
+		foreach ( $processors as $processor ) {
+			if ( $processor instanceof $filter_by_class ) {
+				if ( Processor\Before::class === $filter_by_class ) {
+					self::run_processor( $processor, $data );
+				} elseif ( Processor\After::class === $filter_by_class ) {
+					$result = self::run_processor( $processor, $data );
+					if ( $result ) {
+						$data[1] = $result;
+					}
+				} else {
+					// TODO: error
```

Comment:
Do.

Reply:
> Do.

Are you sure? we said that we want avoiding of throwing exception in this level...